future, loop: use Future concept as future constraints

tchaikov opened the pull request:
- future: use Future concept instead of same_as<future<...>>
- core/loop: use Future concept in do_for_each_impl

avikivity commented on include/seastar/core/future.hh:
```diff
@@ -1834,7 +1834,7 @@ public:
     /// future<>, the handler function does not need to return anything.
     template <typename Func>
-    SEASTAR_CONCEPT( requires ::seastar::InvokeReturns<Func, future<T SEASTAR_ELLIPSIS>, std::exception_ptr>
+    SEASTAR_CONCEPT( requires ::seastar::InvokeReturnsAnyFuture<Func, std::exception_ptr>
```
is_future is not an extension point. I don't mind the change itself, but Seastar cannot be expected to work with non-Seastar futures.

Note, in ScyllaDB we also have futures that return outcomes. This is clearly a useful concept. See https://github.com/scylladb/scylladb/blob/master/utils/result.hh. We may want to make it part of Seastar, perhaps using std::expected.

tchaikov replied:
i see. since the non-Seastar future can be futurized to the expected `seastar::future<T...>`, i will update the change to loosen the constraint so it also allows the types which can be `seastar::futurized()` to the expected `seastar::future<T...>`.

tchaikov replied:
after a second thought, to check the return type `seastar::futurize_invoke(func, args...)` is not better than what we have now. actually, it defeats the purpose of  these compound constraints. the existing constraints explain the requirement in a clear way. while `seastar::futurize_invoke()` is more like the implementation details of how we futurize the return value of `func`. to return outcomes does not address our needs, our customized future is a templated system requiring developer to handle all possible errors using a templated API interface.

avikivity replied:
We have something similar, on top of boost::outcome. 

https://github.com/scylladb/scylladb/blob/master/utils/result_try.hh
https://github.com/scylladb/scylladb/blob/master/utils/result_combinators.hh
https://github.com/scylladb/scylladb/blob/master/utils/result_loop.hh

Copying @piodul the author. I remember errorator now, but I forgot it when we re-developed what's likely very similar.

tchaikov replied:
@avikivity thank you very much. i will take a closer look at the "result" subsystem this weekend. it does looks quite like our own errorator . and it seems it suffers the same pain as errorator does -- we are practically replicating more and more facilities which are supposed to work with the vanilla `seastar::future`. we just mirrored `do_for_each()` for erraror modeling after `seastar::do_for_each()` couple days ago: https://github.com/ceph/ceph/commit/40955f60499323ba1c0ae0bfd200793744e67152 .

avikivity replied:
Well, nobody should use do_for_each in 2022. You should use a coroutine and a regular for loop (an exception is if the loop is expected to only involve ready futures).

But the point stands.

1. exceptions are extremely slow
2. we still need to handle error conditions


Perhaps the solution is to specialize the various future-handling helpers to support future<std::expected<T, E>>.

As a trivial case we'd have

```c++
template <typename Func>
template <typename T, typename E>
requires std::invocable<Func, T>
expected_future_t<futurize_t<std::invoke_result_t<Func, T>>, E>
future<std::expected<T, E>>::then(Func&& func) { ... }
```

If the input future contains an exception or an error, it is propagated. Otherwise Func is invoked and the result is wrapped in std::expected<T, E> and seastar::future.

The alternative is to just support seastar::future<T, E>. This has an efficiency advantage because the existing value/future/exception discriminator can be extended to be value/future/exception/error.

tchaikov replied:
> Well, nobody should use do_for_each in 2022. You should use a coroutine and a regular for loop (an exception is if the loop is expected to only involve ready futures).

agreed. i like C++20 coroutines. it's much easier than C++ sender/receiver and future/promise. but the Ceph project just migrated to C++20 3 months ago. and it took us more time to fix the various fallouts. moreover, we have two systems built on top of `seastar::future`. 

1. errorator. which allows us to peel an onion of `bo::result<T, exception_container<Errors...>` using a more structured way, if i explain this in the scylladb terminology. so at a higher level, the consumer of a certain API only needs to care about a much smaller set of `exception_container` or just a `T`. and in our use case, as you put, we do have the requirement to differentiate error from exception, as sometime, for instance, `ENOENT` can happen quite often even in critical path.
2. cancellation. due to the nature of distributed system and how we model the service, the precondition for serving an RPC might not hold anymore, so we need to cancel an async in-flight RPC call. so we introduced the concept of interruptible future, which allows us to attach an interruptible condition to the future, so whenever the coroutine advances, it checks the condition, and cancel the execution the condition evaluates to true.

> 
> But the point stands.
> 
> 1. exceptions are extremely slow
> 2. we still need to handle error conditions

exactly.

> 
> Perhaps the solution is to specialize the various future-handling helpers to support future<std::expected<T, E>>.
> 
> As a trivial case we'd have
> 
> ```c++
> template <typename Func>
> template <typename T, typename E>
> requires std::invocable<Func, T>
> expected_future_t<futurize_t<std::invoke_result_t<Func, T>>, E>
> future<std::expected<T, E>>::then(Func&& func) { ... }
> ```
> 
> If the input future contains an exception or an error, it is propagated. Otherwise Func is invoked and the result is wrapped in std::expected<T, E> and seastar::future.
> 

what is `expected_future_t<T, E>` ? is it `future<expected<T, E>>` ? if that's the case, instead of returning `future<expected<future<T>, E>>`, i think we should return `future<expected<T, E>>`. an example would look like

```c++
/// assuming input.read() returns future<std::expected<T, E>>
future<std::expected<size_t, error_io>> dd(input_stream& input, output_stream& output) {
  constexpr size_t block_size = 4096;
  // input.read() returns future<std::expected<T, E>>, where
  // T is size_t
  // E is exception_container<error_backoff, error_crc_mismatch, error_disconnected>
  input.read(block_size).handle_error([](error_backoff e) {
    // take a break, and try again?
  }).handle_error([](error_crc_mismatch e) {
    // find another replica
    return make_ready_future<std::expected<size_t, error_io>>(std::unexpected(error_io{}));
  }).handle_error([](error_disconnected e) {
    // use a different transport maybe
    return make_ready_future<std::expected<size_t, error_io>>(std::unexpected(error_io{}));
  }).then([&output](auto buffer) {
    // output.write() returns a future<size_t, error_io>
    return output.write(buffer);
  });
}
```

but it would be difficult to force the caller of the API to "peel" the onion by handling all the possible errors with coroutine. i think this is the inherent limit of `co_await` in comparison to the chained calls -- one cannot force its caller to check all the errors in the `exception_container` in `std::expected`. but i believe we can have a facility to ensure all cases are taken care of as long as developer tries to handle the errors in `exception_container`.

> The alternative is to just support seastar::future<T, E>. This has an efficiency advantage because the existing value/future/exception discriminator can be extended to be value/future/exception/error.

yeah, probably this would be easier if we need to adapt `future<T, E>` to seastar primitives. for instance, most (if not all) of them should ensure that if the callee coroutine fails, the `E` should be propagated back to the caller.

avikivity replied:
> > Well, nobody should use do_for_each in 2022. You should use a coroutine and a regular for loop (an exception is if the loop is expected to only involve ready futures).
> 
> agreed. i like C++20 coroutines. it's much easier to reason about than C++ sender/receiver and future/promise. but the Ceph project just migrated to C++20 3 months ago. and it took us more time to fix the various fallouts. moreover, we have two systems built on top of `seastar::future`.
> 
> 1. errorator. which allows us to peel an onion of `bo::result<T, exception_container<Errors...>` using a more structured way, if i explain this in the scylladb terminology. so at a higher level, the consumer of a certain API only needs to care about a much smaller set of `exception_container` or just a `T`. and in our use case, as you put, we do have the requirement to differentiate error from exception, as sometime, for instance, `ENOENT` can happen quite often even in critical path.
> 2. cancellation. due to the nature of distributed system and how we model the service, the precondition for serving an RPC might not hold anymore, so we need to cancel an async in-flight RPC call. so we introduced the concept of interruptible future, which allows us to attach an interruptible condition to the future, so whenever the coroutine advances, it checks the condition, and cancel the execution the condition evaluates to true. i think we can address this by defining yet another `future` and to define `await_transform` to conditionlize on the type of transformed expression, so we can check the cancellation token, and propagate it into the cancellable future when the coroutine state machine advances. anyway, this is quite a different topic. i will try to start a discussion in the discussion group for more inputs.

I tried before to create a template coroutine wrapper, similar to std::coroutine_traits, that will allow the user to connect their own custom type with a seastar::task (without exposing seastar::task), so they user can impose whatever rules they like and still get seastar to schedule their coroutine.


> 
> > But the point stands.
> > 
> > 1. exceptions are extremely slow
> > 2. we still need to handle error conditions
> 
> exactly.
> 
> > Perhaps the solution is to specialize the various future-handling helpers to support future<std::expected<T, E>>.
> > As a trivial case we'd have
> > ```c++
> > template <typename Func>
> > template <typename T, typename E>
> > requires std::invocable<Func, T>
> > expected_future_t<futurize_t<std::invoke_result_t<Func, T>>, E>
> > future<std::expected<T, E>>::then(Func&& func) { ... }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If the input future contains an exception or an error, it is propagated. Otherwise Func is invoked and the result is wrapped in std::expected<T, E> and seastar::future.
> 
> what is `expected_future_t<T, E>` ? is it `future<expected<T, E>>` ? if that's the case, instead of returning `future<expected<future<T>, E>>`, i think we should return `future<expected<T, E>>`. an example would look like


I think that `future<T, E>` should behave like `std::expected<T, E>` but with future support, and we should provide conversions to and from std::expected.

> 
> ```c++
> /// assuming input.read() returns future<std::expected<T, E>>
> future<std::expected<size_t, error_io>> dd(input_stream& input, output_stream& output) {
>   constexpr size_t block_size = 4096;
>   // input.read() returns future<std::expected<T, E>>, where
>   // T is size_t
>   // E is exception_container<error_backoff, error_crc_mismatch, error_disconnected>
>   input.read(block_size).handle_error([](error_backoff e) {
>     // take a break, and try again?
>   }).handle_error([](error_crc_mismatch e) {
>     // find another replica
>     return make_ready_future<std::expected<size_t, error_io>>(std::unexpected(error_io{}));
>   }).handle_error([](error_disconnected e) {
>     // use a different transport maybe
>     return make_ready_future<std::expected<size_t, error_io>>(std::unexpected(error_io{}));
>   }).then([&output](auto buffer) {
>     // output.write() returns a future<size_t, error_io>
>     return output.write(buffer);
>   });
> }
> ```
> 

How about

```c++
    auto e = co_await input.read(block_size);
    // use std::expected utilities to handle errors
```

All those extra continuations are just terrible.

> but it would be difficult to force the caller of the API to "peel" the onion by handling all the possible errors with coroutine. i think this is the inherent limit of `co_await` in comparison to the chained calls -- one cannot force its caller to check all the errors in the `exception_container` in `std::expected`. but i believe we can have a facility to ensure all cases are taken care of as long as developer tries to handle the errors in `exception_container`.

That's a problem for std::expected's developer to solve. Seastar can't solve it.

> 
> > The alternative is to just support seastar::future<T, E>. This has an efficiency advantage because the existing value/future/exception discriminator can be extended to be value/future/exception/error.
> 
> yeah, probably this would be easier if we need to adapt `future<T, E>` to seastar primitives. for instance, most (if not all) of them should ensure that if the callee coroutine fails, the `E` should be propagated back to the caller.

We could do that with `future<std::expected<T, E>>` too, by specializing stuff.

tchaikov commented:
changelog

- dropped the `InvokeReturnsAnyFuture<>` change.
- added change to replace `is_future<>` with `Future<>`
- added change to replace `i++` requirement with named `std::weakly_incrementable<>` concept
- fix the `i++` requirement. it should be `++i`.
